Add section on managing abstraction v complexity (#98)
sourcegraph/handbook: content/direction/code-graph/index.md

> We want to make using Sourcegraph to understand and modify code so easy that anyone can do it, including people not traditionally considered developers. We want to [democratize coding](../../company/strategy.md#10-year-vision) by making it so that everyone in the company can engage with the organization's code, safely making changes to implement their department's priorities and interests, without need for an intermediary to do it for them.

## Principles

### Abstraction vs. complexity

There are two ways to think about categories of developer tools or features today:

- Abstraction tools, which simplify tasks away by providing building blocks, sane defaults, and frameworks to build on. Most existing developer tools are of this variety.

> - Complexity-exploring tools, which help with complex problems such as finding and fixing issues in an existing, heterogeneous software system.. one that was created using the existing tools developers use today, with different languages, frameworks, and versions in constant flux, all interacting with each other.

Reviewer: typo in "system.. one"?

Author: ```suggestion
- Complexity-exploring tools, which help with complex problems such as finding and fixing issues in existing, heterogeneous software systems. I.e., ones that were created using the existing tools developers use today, with different languages, frameworks, and versions in constant flux, all interacting with each other.
```

This emerging need for complexity-exploring tooling is one we see Sourcegraph playing an important role in. The reality is that software tech stacks today look more like a rainforest — with animals and plants co-existing, competing, living, dying, growing, interacting in unplanned ways — than like a planned garden, and we are mindfully designing our the Code Graph to let you embrace this complexity rather than hiding it, or pushing you towards a specific single solution.

Practically, this means a few things as we think about how the Code Graph should work:

> 1. We can find and index your code wherever you host it or whatever language it is written in because we fit into your existing ecosystem.

Reviewer 2: ```suggestion
1. Sourcegraph fits in to your existing ecosystem – wherever you host your code, whatever language it's in, or however many repos you use. 
```
I've heard from others that we used to explicitly and do still try to uphold "treat monorepo and multi-repo codebases the same." This gets tricky (sometimes extremely tricky, in the case of code insights), but it's something that makes sense to me. If we agree and want to keep this going forward, I'd love to document that here somehow a bit more directly. 

I might also suggest changing "find and index" to be a bit broader, (is batch changes find but also "changing/acting/replacing"? Is code insights also "tracking"?) but don't have as strong a preference to that point

Author: Really great improvement, thanks @Joelkw

1. By surfacing relationships (not just code itself, but data types, configution, and more) we help you build a clear, meaningful model of your software.

> 1. We make this information easy to explore, with the most relevant information up front, easy to understand, and easy to share by revealing complexity as it is needed.

Reviewer 2: ```suggestion
1. We make this information easy to explore by revealing complexity as needed, with the most relevant information up front, easy to understand, and easy to share. 
```
(just fixing misplaced modifier)
